loopy/target/c/codegen/expression.py

```diff
@@ -711,23 +717,43 @@ def base_impl(expr, type_context):
                 elif is_zero(expr.exponent - 2):
                     return self.rec(expr.base*expr.base, type_context)
 
-            return type(expr)(
-                    self.rec(expr.base, type_context),
-                    self.rec(expr.exponent, type_context))
+            if exponent_dtype.is_integral():
+                from loopy.codegen import SeenFunction
+                func_name = ("loopy_pow_"
+                        f"{base_dtype.numpy_dtype}_{exponent_dtype.numpy_dtype}")
+
+                self.codegen_state.seen_functions.add(
+                        SeenFunction(
+                            "int_pow", func_name,
+                            (base_dtype, exponent_dtype),
+                            (tgt_dtype, )))
+                return var("loopy_pow_"
+                        f"{base_dtype.numpy_dtype}_{exponent_dtype.numpy_dtype}")(
+                                self.rec(expr.base), self.rec(expr.exponent))
+            else:
+                from loopy.types import to_loopy_type
+                loopy_f64_dtype = to_loopy_type(np.float64,
+                        target=self.kernel.target)
+                return self.wrap_in_typecast(
+                        loopy_f64_dtype,
+                        tgt_dtype,
+                        var("pow")(self.rec(expr.base, type_context,
+                                            loopy_f64_dtype),
+                                   self.rec(expr.exponent, type_context,
+                                            loopy_f64_dtype)))
```

Comment:
This is not correct for (at least) OpenCL, which offers f32 `pow`.

Reply:
Agreed and fixed.